sqs/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/SqsSink.scala

```diff
@@ -0,0 +1,335 @@
+      * Even if the incoming event flow dries up, all buffered events will eventually get sent.
+      * Whenever TimeThreshold milliseconds have passed since the last call to flush, call flush.
+      * @param interval When to schedule the next flush
+      */
+    def scheduleFlush(interval: Long = TimeThreshold): Unit = {
+      executorService.schedule(
+        new Thread {
+          override def run(): Unit = {
+            val lastFlushed = getLastFlushTime()
+            val currentTime = System.currentTimeMillis()
+            if (currentTime - lastFlushed >= TimeThreshold) {
+              flush()
+              scheduleFlush(TimeThreshold)
+            } else {
+              scheduleFlush(TimeThreshold + lastFlushed - currentTime)
+            }
+          }
+        },
+        interval,
+        MILLISECONDS
+      )
+      ()
+    }
+  }
+
+  def sendBatch(batch: List[EventBuffer.Event], nextBackoff: Long = minBackoff): Unit =
+    if (batch.nonEmpty) {
+      log.info(s"Writing ${batch.size} Thrift records to SQS queue ${queueName}")
+
+      multiPut(batch).onComplete {
+        case Success(s) =>
+          log.info(s"Successfully wrote ${batch.size - s.size} out of ${batch.size} messages.")
+          if (s.nonEmpty) {
+            s.foreach { f =>
+              log.error(
+                s"Message failed with error code [${f._2.code}] and message [${f._2.message}]"
+              )
+            }
+            log.error(s"Retrying in $nextBackoff milliseconds...")
+            scheduleBatch(s.keys.toList, nextBackoff)
+          }
+        case Failure(f) =>
+          log.error("Writing failed.", f)
+          log.error(s"Retrying in $nextBackoff milliseconds...")
+          scheduleBatch(batch, nextBackoff)
+      }
+    }
+
+  // The batch request can return unsuccessful entries. In this case, we return a map of Events to be retried and the reasons they failed to be inserted.
+  def multiPut(batch: List[EventBuffer.Event]): Future[Map[EventBuffer.Event, BatchResultErrorInfo]] =
+    Future {
+      val MaxSqsBatchSize = 10 // SQS limit
+      batch
+        .map { e =>
+          (e, toSqsMessage(e))
+        }
+        .toMap
+        .grouped(MaxSqsBatchSize)
+        .flatMap { msgGroup =>
+          val batchRequest =
+            new SendMessageBatchRequest().withQueueUrl(queueName).withEntries(msgGroup.values.toList.asJava)
+          val response = client.sendMessageBatch(batchRequest)
+          val failures = response
+            .getFailed
+            .asScala
+            .toList
+            .map { bree =>
+              (bree.getId, BatchResultErrorInfo(bree.getCode, bree.getMessage))
+            }
+            .toMap
+          // Events to retry and reasons for failure
+          msgGroup.collect {
+            case (e, m) if (failures.keys.toSet.contains(m.getId)) =>
+              (e, failures.getOrElse(m.getId, BatchResultErrorInfo("unknown code", "unknown message")))
```

Comment:
```suggestion
              (e, failures(m.getId))
```
Safe thanks to the test before